### PR TITLE
Amazon: Add to cart functionality

### PR DESCRIPTION
**Pull request by nareshmahendra**
**Flow/Test steps :**
Launch amazon.in
search for 'iphone 13'
validate the results
click on 1st product
validate the product details page and the product title
validate add to cart
click on add to cart
go to cart
validate the same product is in the cart

**Changes :** 
_A. Added:_
 1. amazon_config.properties in config package
 2. amazon_cart.featues in e2etest features package
 3. 5 Bls classes in new amazon BLs packages
 4. 3 Step definations in amazon step defination package
 5. 5 Abstract classes & their implementation in amazon screens package

_B.Edit_
 1. gradle.build
 2.environment.json file in resources package

### CODE REVIEW

File: src/test/java/com/znsio/sample/e2e/screen/amazon/NavBarScreen.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                .getId());
+        Platform platform = Runner.fetchPlatform(Thread.currentThread()
+                .getId());
+        LOGGER.info(SCREEN_NAME + ": Driver type: " + driver.getType() + ": Platform: " + platform);
+        Visual visually = fetchEyes(Thread.currentThread()
+                .getId());
+
+        switch (platform) {
+            case web:
+                return new NavBarWeb(driver, visually);
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/NavBarScreen.java:31]: Class name should be NavBarScreenWeb

File: src/test/java/com/znsio/sample/e2e/screen/amazon/ProductDetailsScreen.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import com.znsio.sample.e2e.screen.amazon.web.ProductDetailsWeb;
+import com.znsio.sample.e2e.screen.theapp.AppLaunchScreen;
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.log4j.Logger;
+
+import static com.znsio.e2e.runner.Runner.fetchDriver;
+import static com.znsio.e2e.runner.Runner.fetchEyes;
+
+public abstract class ProductDetailsScreen {
+    private static final String SCREEN_NAME = AppLaunchScreen.class.getSimpleName();
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/ProductDetailsScreen.java:16]: class name

File: src/test/java/com/znsio/sample/e2e/screen/amazon/ProductDetailsScreen.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+                .getId());
+        Platform platform = Runner.fetchPlatform(Thread.currentThread()
+                .getId());
+        LOGGER.info(SCREEN_NAME + ": Driver type: " + driver.getType() + ": Platform: " + platform);
+        Visual visually = fetchEyes(Thread.currentThread()
+                .getId());
+
+        switch (platform) {
+            case web:
+                return new ProductDetailsWeb(driver, visually);
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/ProductDetailsScreen.java:30]: ProductDetailsScreenWeb

File: src/test/java/com/znsio/sample/e2e/screen/amazon/SearchResultsScreen.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import org.apache.log4j.Logger;
+
+import java.util.List;
+
+import static com.znsio.e2e.runner.Runner.fetchDriver;
+import static com.znsio.e2e.runner.Runner.fetchEyes;
+
+public abstract class SearchResultsScreen {
+
+    private static final String SCREEN_NAME = AppLaunchScreen.class.getSimpleName();
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/SearchResultsScreen.java:19]: Class name

File: src/test/java/com/znsio/sample/e2e/screen/amazon/SearchResultsScreen.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                .getId());
+        Platform platform = Runner.fetchPlatform(Thread.currentThread()
+                .getId());
+        LOGGER.info(SCREEN_NAME + ": Driver type: " + driver.getType() + ": Platform: " + platform);
+        Visual visually = fetchEyes(Thread.currentThread()
+                .getId());
+
+        switch (platform) {
+            case web:
+                return new SearchResultsWeb(driver, visually);
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/SearchResultsScreen.java:33]: SearchResultsScreenWeb

File: src/test/java/com/znsio/sample/e2e/screen/amazon/ShoppingCartScreen.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import com.znsio.sample.e2e.screen.amazon.web.ShoppingCartWeb;
+import com.znsio.sample.e2e.screen.theapp.AppLaunchScreen;
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.log4j.Logger;
+
+import static com.znsio.e2e.runner.Runner.fetchDriver;
+import static com.znsio.e2e.runner.Runner.fetchEyes;
+
+public abstract class ShoppingCartScreen {
+    private static final String SCREEN_NAME = AppLaunchScreen.class.getSimpleName();
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/ShoppingCartScreen.java:16]: class name

File: src/test/java/com/znsio/sample/e2e/screen/amazon/ShoppingCartScreen.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+                .getId());
+        Platform platform = Runner.fetchPlatform(Thread.currentThread()
+                .getId());
+        LOGGER.info(SCREEN_NAME + ": Driver type: " + driver.getType() + ": Platform: " + platform);
+        Visual visually = fetchEyes(Thread.currentThread()
+                .getId());
+
+        switch (platform) {
+            case web:
+                return new ShoppingCartWeb(driver, visually);
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/ShoppingCartScreen.java:31]: ShoppingCartScreenWeb

File: src/test/java/com/znsio/sample/e2e/screen/amazon/ShoppingCartScreen.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+            default:
+                throw new NotImplementedException(SCREEN_NAME + " is not implemented in " + Runner.platform);
+        }
+    }
+
+    public abstract String getProductTitle();
+
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/ShoppingCartScreen.java:38]: remove extra empty lines

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/HomePageWeb.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+import com.znsio.e2e.tools.Driver;
+import com.znsio.e2e.tools.Visual;
+import com.znsio.sample.e2e.screen.amazon.LaunchScreen;
+import com.znsio.sample.e2e.screen.theapp.AppLaunchScreen;
+import com.znsio.sample.e2e.screen.web.theapp.AppLaunchScreenWeb;
+import org.apache.log4j.Logger;
+
+public class HomePageWeb extends LaunchScreen {
+
+    private static final String SCREEN_NAME = AppLaunchScreen.class.getSimpleName();
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/web/HomePageWeb.java:12]: class name

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/ProductDetailsWeb.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+import com.znsio.e2e.tools.Driver;
+import com.znsio.e2e.tools.Visual;
+import com.znsio.sample.e2e.screen.amazon.ProductDetailsScreen;
+import com.znsio.sample.e2e.screen.theapp.AppLaunchScreen;
+import com.znsio.sample.e2e.screen.web.theapp.AppLaunchScreenWeb;
+import org.apache.log4j.Logger;
+import org.openqa.selenium.By;
+
+public class ProductDetailsWeb extends ProductDetailsScreen {
+    private static final String SCREEN_NAME = AppLaunchScreen.class.getSimpleName();
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/web/ProductDetailsWeb.java:12]: class name

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/ProductDetailsWeb.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    private static final Logger LOGGER = Logger.getLogger(AppLaunchScreenWeb.class.getName());
+    private final Driver driver;
+    private final Visual visually;
+
+    private final By productTitle = By.cssSelector("span#productTitle");
+
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/web/ProductDetailsWeb.java:18]: remove extra empty lines

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/SearchResultsWeb.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    public SearchResultsWeb(Driver driver, Visual visually) {
+        this.driver = driver;
+        this.visually = visually;
+    }
+
+    /**
+     * Opens product by {@code index} .
+     */
+    @Override
+    public String openProductByIndex(int index) {
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/web/SearchResultsWeb.java:33]: are you opening a product or selecting a product?
What is index?

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/SearchResultsWeb.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+        visually.takeScreenshot(SCREEN_NAME, " Product details page ");
+        return productName;
+    }
+
+    /**
+     * Swiches to new tab .
+     */
+    private void switchToNewTab() {
+        String parent = driver.getInnerDriver().getWindowHandle();
+        driver.getInnerDriver().switchTo().window(driver.getInnerDriver().getWindowHandles().stream().filter(s -> !s.equals(parent)).findFirst().get());
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/web/SearchResultsWeb.java:50]: I believe driver should have a method to switch to a window. Can you check and confirm?

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/ShoppingCartWeb.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+import com.znsio.e2e.tools.Driver;
+import com.znsio.e2e.tools.Visual;
+import com.znsio.sample.e2e.screen.amazon.ShoppingCartScreen;
+import com.znsio.sample.e2e.screen.theapp.AppLaunchScreen;
+import com.znsio.sample.e2e.screen.web.theapp.AppLaunchScreenWeb;
+import org.apache.log4j.Logger;
+import org.openqa.selenium.By;
+
+public class ShoppingCartWeb extends ShoppingCartScreen {
+    private static final String SCREEN_NAME = AppLaunchScreen.class.getSimpleName();
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/web/ShoppingCartWeb.java:12]: class name

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/pageFragment/NavBarWeb.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+import com.znsio.sample.e2e.screen.theapp.AppLaunchScreen;
+import com.znsio.sample.e2e.screen.web.theapp.AppLaunchScreenWeb;
+import org.apache.log4j.Logger;
+import org.openqa.selenium.By;
+
+public class NavBarWeb extends NavBarScreen {
+
+    public static final By searchInput = By.cssSelector("input#twotabsearchtextbox");
+    public static final By itemsIncart = By.cssSelector("span#nav-cart-count");
+    private static final String SCREEN_NAME = AppLaunchScreen.class.getSimpleName();
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/web/pageFragment/NavBarWeb.java:16]: class name

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/pageFragment/NavBarWeb.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    private final Visual visually;
+    private final By searchButton = By.cssSelector("span#nav-search-submit-text");
+
+    public NavBarWeb(Driver driver, Visual visually) {
+        this.driver = driver;
+        this.visually = visually;
+        visually.takeScreenshot(SCREEN_NAME, "Home screen");
+    }
+
+    public By getSearchInput() {
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/web/pageFragment/NavBarWeb.java:28]: method name is not clear what is exactly being done

File: src/test/java/com/znsio/sample/e2e/steps/amazon/ProductDetailsSteps.java [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+import com.znsio.sample.e2e.businessLayer.amazon.NavBarBL;
+import com.znsio.sample.e2e.businessLayer.amazon.ProductDetailsBL;
+import com.znsio.sample.e2e.entities.SAMPLE_TEST_CONTEXT;
+import com.znsio.sample.e2e.steps.TheAppSteps;
+import io.cucumber.java.en.And;
+import io.cucumber.java.en.Then;
+import org.apache.log4j.Logger;
+
+public class ProductDetailsSteps {
+    private static final Logger LOGGER = Logger.getLogger(TheAppSteps.class.getName());
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/steps/amazon/ProductDetailsSteps.java:16]: class name

File: src/test/java/com/znsio/sample/e2e/steps/amazon/ProductSearchSteps.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+import com.znsio.sample.e2e.businessLayer.amazon.LaunchScreenBL;
+import com.znsio.sample.e2e.businessLayer.amazon.SearchResultsBL;
+import com.znsio.sample.e2e.entities.SAMPLE_TEST_CONTEXT;
+import com.znsio.sample.e2e.steps.TheAppSteps;
+import io.cucumber.java.en.And;
+import io.cucumber.java.en.Given;
+import org.apache.log4j.Logger;
+
+public class ProductSearchSteps {
+    private static final Logger LOGGER = Logger.getLogger(TheAppSteps.class.getName());
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/steps/amazon/ProductSearchSteps.java:16]: class name

File: src/test/java/com/znsio/sample/e2e/steps/amazon/ProductSearchSteps.java [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+                Runner.platform));
+        new LaunchScreenBL(SAMPLE_TEST_CONTEXT.ME, Runner.platform).searchProduct(productName).verifySearchResults(productName);
+    }
+
+
+    @And("I Open first search result details")
+    public void iOpenFirstSearchResultDetails() {
+        LOGGER.info(System.out.printf("I search for product, Platform: '%s'", SAMPLE_TEST_CONTEXT.ME,
+                Runner.platform));
+        new SearchResultsBL(SAMPLE_TEST_CONTEXT.ME, Runner.platform).openFirstProduct().verifyProductName();
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/steps/amazon/ProductSearchSteps.java:48]: openFirstProduct should automatically verify if the business operation was done correctly, or not

File: src/test/java/com/znsio/sample/e2e/steps/amazon/ShoppingCartSteps.java [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+import com.znsio.e2e.runner.Runner;
+import com.znsio.e2e.tools.Drivers;
+import com.znsio.sample.e2e.businessLayer.amazon.ShoppingCartBL;
+import com.znsio.sample.e2e.entities.SAMPLE_TEST_CONTEXT;
+import com.znsio.sample.e2e.steps.TheAppSteps;
+import io.cucumber.java.en.Then;
+import org.apache.log4j.Logger;
+
+public class ShoppingCartSteps {
+    private static final Logger LOGGER = Logger.getLogger(TheAppSteps.class.getName());
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/steps/amazon/ShoppingCartSteps.java:14]: class name

File: src/test/java/com/znsio/sample/e2e/steps/amazon/ShoppingCartSteps.java [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+                .getId());
+        LOGGER.info("context: " + context.getTestName());
+        allDrivers = (Drivers) context.getTestState(SAMPLE_TEST_CONTEXT.ALL_DRIVERS);
+        LOGGER.info("allDrivers: " + (null == allDrivers));
+    }
+
+    @Then("I verify the cart details")
+    public void iVerifyTheCartDetails() {
+        LOGGER.info(String.format("I verify the cart details, Platform: '%s'", SAMPLE_TEST_CONTEXT.ME,
+                Runner.platform));
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/steps/amazon/ShoppingCartSteps.java:29]: why is this step empty?

File: src/test/resources/com/znsio/sample/e2e/features/amazon_cart.feature [hunk 22]
```diff
@@ -0,0 +1,10 @@
+@amazon
+Feature: Amazon
+
+  @web @amazon @amazon-test
+  Scenario: Add to cart
+    Given I launch amazon application
+    And I search for "Iphone 13" and verify results
```
Comment (anandbagmar) [inline src/test/resources/com/znsio/sample/e2e/features/amazon_cart.feature:7]: Change
`And I search for "Iphone 13" and verify results`
to
`And I search for "Iphone 13"` --> this should automatically do the verification

File: src/test/resources/com/znsio/sample/e2e/features/amazon_cart.feature [hunk 23]
```diff
@@ -0,0 +1,10 @@
...
+    When I Open first search result details
```
Comment (anandbagmar) [inline src/test/resources/com/znsio/sample/e2e/features/amazon_cart.feature:8]: This should have been `When I open first search result details`

`When I Open first search result details` -> what does "open first search result details" mean?

File: src/test/resources/com/znsio/sample/e2e/features/amazon_cart.feature [hunk 24]
```diff
@@ -0,0 +1,10 @@
...
+    And I add above item to cart
+    Then I verify details in cart
```
Comment (anandbagmar) [inline src/test/resources/com/znsio/sample/e2e/features/amazon_cart.feature:10]: Verification step is not required. It should be done automatically as part of the earlier step

File: src/test/resources/environments.json [hunk 25]
```diff
@@ -1,8 +1,9 @@
 {
   "prod": {
     "BASE_URL_API": "http://the-internet.herokuapp.com",
-    "BASE_URL": "http://the-internet.herokuapp.com/",
+    "BASE_URL": "https://www.amazon.in/",
     "IMAGES_BASE_URL": "http://images.google.com/",
-    "BING_BASE_URL": "http://bing.com/images"
+    "BING_BASE_URL": "http://bing.com/images",
+    "BROWSER": "firefox"
```
Comment (anandbagmar) [inline src/test/resources/environments.json:7]: why is BROWSER defined here? It is not picked up from here. Delete it

Comment (nareshmahendra) [reply src/test/resources/environments.json:7]: Acknowledge

File: src/test/java/com/znsio/sample/e2e/screen/amazon/LaunchScreen.java [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+                .getId());
+        Platform platform = Runner.fetchPlatform(Thread.currentThread()
+                .getId());
+        LOGGER.info(SCREEN_NAME + ": Driver type: " + driver.getType() + ": Platform: " + platform);
+        Visual visually = fetchEyes(Thread.currentThread()
+                .getId());
+
+        switch (platform) {
+            case web:
+                return new HomePageWeb(driver, visually);
```
[the remark this answers is not in the document]
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/LaunchScreen.java:33]: Changed "HomePageWeb" to LaunchPageWeb

File: src/test/java/com/znsio/sample/e2e/steps/amazon/ShoppingCartSteps.java (hunk 21 above)
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/steps/amazon/ShoppingCartSteps.java:29]: Removed step

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/ProductDetailsWeb.java [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+     */
+    public String getProductName() {
+        return driver.findElement(productTitle).getText();
+    }
+
+
+    /**
+     * Clicks add to cart button .
+     */
+    public ProductDetailsWeb clickAddToCardButton() {
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/web/ProductDetailsWeb.java:41]: return ProductDetailsScreen.get();

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/ProductDetailsWeb.java [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    /**
+     * Clicks view cart button .
+     */
+    public ShoppingCartWeb clickViewCartButton() {
+        visually.takeScreenshot(SCREEN_NAME, " Before moving to cart page");
+        driver.waitForClickabilityOf(viewCart);
+        driver.findElement(viewCart).click();
+        return new ShoppingCartWeb(driver, visually);
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/amazon/web/ProductDetailsWeb.java:55]: return ShoppingCartScreen.get();

File: src/test/resources/com/znsio/sample/e2e/features/amazon_cart.feature (hunk 23 above)
Comment (anandbagmar) [reply src/test/resources/com/znsio/sample/e2e/features/amazon_cart.feature:8]: ex: `When I select the first product from the search results`

Comment (nareshmahendra) [reply src/test/resources/com/znsio/sample/e2e/features/amazon_cart.feature:8]: Done

File: src/test/resources/com/znsio/sample/e2e/features/amazon_cart.feature (hunk 24 above)
Comment (nareshmahendra) [reply src/test/resources/com/znsio/sample/e2e/features/amazon_cart.feature:10]: Acknowledge

File: src/test/java/com/znsio/sample/e2e/steps/amazon/ProductDetailsSteps.java (hunk 17 above)
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/steps/amazon/ProductDetailsSteps.java:16]: Done

File: src/test/java/com/znsio/sample/e2e/steps/amazon/ProductSearchSteps.java (hunk 18 above)
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/steps/amazon/ProductSearchSteps.java:16]: Done

File: src/test/java/com/znsio/sample/e2e/steps/amazon/ProductSearchSteps.java (hunk 19 above)
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/steps/amazon/ProductSearchSteps.java:48]: Done

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/pageFragment/NavBarWeb.java (hunk 16 above)
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/web/pageFragment/NavBarWeb.java:28]: Changed to getSearchInputByLocator

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/HomePageWeb.java (hunk 9 above)
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/web/HomePageWeb.java:12]: Done

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/SearchResultsWeb.java [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+import com.znsio.sample.e2e.screen.web.theapp.AppLaunchScreenWeb;
+import org.apache.log4j.Logger;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class SearchResultsWeb extends SearchResultsScreen {
+    private static final String SCREEN_NAME = AppLaunchScreen.class.getSimpleName();
```
[the remark this answers is not in the document]
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/web/SearchResultsWeb.java:17]: Done

File: src/test/java/com/znsio/sample/e2e/screen/amazon/web/SearchResultsWeb.java (hunk 12 above)
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/web/SearchResultsWeb.java:33]: Changed to productIndexInSearchResults

File: src/test/java/com/znsio/sample/e2e/screen/amazon/LaunchScreen.java [hunk 30]
```diff
@@ -0,0 +1,41 @@
...
+import com.znsio.sample.e2e.screen.theapp.AppLaunchScreen;
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.log4j.Logger;
+
+import static com.znsio.e2e.runner.Runner.fetchDriver;
+import static com.znsio.e2e.runner.Runner.fetchEyes;
+
+public abstract class LaunchScreen {
+
+    private static final String SCREEN_NAME = AppLaunchScreen.class.getSimpleName();
```
[the remark this answers is not in the document]
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/LaunchScreen.java:17]: Done

File: src/test/java/com/znsio/sample/e2e/screen/amazon/LaunchScreen.java [hunk 31]
```diff
@@ -0,0 +1,41 @@
...
+    private static final Logger LOGGER = Logger.getLogger(SCREEN_NAME);
+
+    private static final String PAGE_TITLE = "Online Shopping site in India: Shop Online for Mobiles, Books, Watches, Shoes and More - Amazon.in";
```
[the remark this answers is not in the document]
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/LaunchScreen.java:20]: Acknowledge

File: src/test/java/com/znsio/sample/e2e/screen/amazon/NavBarScreen.java [hunk 32]
```diff
@@ -0,0 +1,41 @@
...
+import com.znsio.sample.e2e.screen.amazon.web.pageFragment.NavBarWeb;
+import com.znsio.sample.e2e.screen.theapp.AppLaunchScreen;
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.log4j.Logger;
+
+import static com.znsio.e2e.runner.Runner.fetchDriver;
+import static com.znsio.e2e.runner.Runner.fetchEyes;
+
+public abstract class NavBarScreen {
+    private static final String SCREEN_NAME = AppLaunchScreen.class.getSimpleName();
```
[the remark this answers is not in the document]
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/NavBarScreen.java:16]: Done

File: src/test/java/com/znsio/sample/e2e/screen/amazon/NavBarScreen.java (hunk 1 above)
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/NavBarScreen.java:31]: Done

File: src/test/java/com/znsio/sample/e2e/screen/amazon/ProductDetailsScreen.java (hunk 2 above)
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/ProductDetailsScreen.java:16]: Done

File: src/test/java/com/znsio/sample/e2e/screen/amazon/SearchResultsScreen.java (hunk 4 above)
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/SearchResultsScreen.java:19]: Done

File: src/test/java/com/znsio/sample/e2e/screen/amazon/ShoppingCartScreen.java (hunk 6 above)
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/ShoppingCartScreen.java:16]: Done

File: src/test/java/com/znsio/sample/e2e/screen/amazon/ShoppingCartScreen.java (hunk 7 above)
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/screen/amazon/ShoppingCartScreen.java:31]: Done

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/LaunchScreenBL.java [hunk 33]
```diff
@@ -0,0 +1,44 @@
...
+
+    /**
+     * Verifies the page title.
+     */
+    public LaunchScreenBL verifyHomePageTitle() throws Exception {
+        softly.assertThat(LaunchScreen.get().verifyPageTitle()).as("Verify home page title").isTrue();
+        return this;
+    }
+
+    /**
```
[the remark this answers is not in the document]
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/businessLayer/amazon/LaunchScreenBL.java:36]: Removed

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/NavBarBL.java [hunk 34]
```diff
@@ -0,0 +1,40 @@
...
+import com.znsio.e2e.runner.Runner;
+import com.znsio.sample.e2e.businessLayer.theapp.AppBL;
+import com.znsio.sample.e2e.screen.amazon.NavBarScreen;
+import org.apache.log4j.Logger;
+import org.assertj.core.api.SoftAssertions;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+public class NavBarBL {
+    private static final Logger LOGGER = Logger.getLogger(AppBL.class.getName());
```
[the remark this answers is not in the document]
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/businessLayer/amazon/NavBarBL.java:14]: Done

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/ProductDetailsBL.java [hunk 35]
```diff
@@ -0,0 +1,49 @@
...
+import com.znsio.e2e.runner.Runner;
+import com.znsio.sample.e2e.businessLayer.theapp.AppBL;
+import com.znsio.sample.e2e.screen.amazon.ProductDetailsScreen;
+import org.apache.log4j.Logger;
+import org.assertj.core.api.SoftAssertions;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+public class ProductDetailsBL {
+    private static final Logger LOGGER = Logger.getLogger(AppBL.class.getName());
```
[the remark this answers is not in the document]
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/businessLayer/amazon/ProductDetailsBL.java:14]: Done

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/ProductDetailsBL.java [hunk 36]
```diff
@@ -0,0 +1,49 @@
...
+        long threadId = Thread.currentThread()
+                .getId();
+        this.context = Runner.getTestExecutionContext(threadId);
+        softly = Runner.getSoftAssertion(threadId);
+        this.currentUserPersona = userPersona;
+        this.currentPlatform = forPlatform;
+        Runner.setCurrentDriverForUser(userPersona, forPlatform, context);
+    }
+
+    /**
```
[the remark this answers is not in the document]
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/businessLayer/amazon/ProductDetailsBL.java:30]: Done

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/SearchResultsBL.java [hunk 37]
```diff
@@ -0,0 +1,56 @@
...
+import com.znsio.sample.e2e.screen.amazon.SearchResultsScreen;
+import org.apache.log4j.Logger;
+import org.assertj.core.api.SoftAssertions;
+
+import java.util.List;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+public class SearchResultsBL {
+
```
[the remark this answers is not in the document]
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/businessLayer/amazon/SearchResultsBL.java:16]: Done

File: src/test/java/com/znsio/sample/e2e/businessLayer/amazon/SearchResultsBL.java [hunk 38]
```diff
@@ -0,0 +1,56 @@
...
+
+    private static final Logger LOGGER = Logger.getLogger(AppBL.class.getName());
```
[the remark this answers is not in the document]
Comment (nareshmahendra) [reply src/test/java/com/znsio/sample/e2e/businessLayer/amazon/SearchResultsBL.java:18]: Done

**Pull request closed without merging**